src/Controller/Component/PaginatorComponent.php

```diff
@@ -187,9 +187,10 @@ public function paginate($object, array $settings = [])
             $query->applyOptions($options);
         }
 
+        $cleanQuery = clone $query;
         $results = $query->all();
         $numResults = count($results);
-        $count = $numResults ? $query->count() : 0;
+        $count = $numResults ? $cleanQuery->count() : 0;
```

Comment:
I had considered changing the EagerLoader to separate 'runtime' state from the configuration state. It seems like a riskier job that might not be doable.